jit/tests/common.rs

```diff
@@ -0,0 +1,203 @@
+}
+
+impl StackMachine {
+    pub fn new() -> StackMachine {
+        StackMachine {
+            stack: Vec::new(),
+            locals: HashMap::new(),
+        }
+    }
+
+    pub fn run(&mut self, code: CodeObject) {
+        for instruction in code.instructions {
+            if self.process_instruction(instruction) {
+                break;
+            }
+        }
+    }
+
+    fn process_instruction(&mut self, instruction: Instruction) -> bool {
+        match instruction {
+            Instruction::LoadConst { value } => self.stack.push(value.into()),
+            Instruction::LoadName {
+                name,
+                scope: NameScope::Free,
+            } => self.stack.push(StackValue::String(name)),
+            Instruction::StoreName { name, .. } => {
+                self.locals.insert(name, self.stack.pop().unwrap());
+            }
+            Instruction::StoreAttr { .. } => {
+                // Do nothing except throw away the stack values
+                self.stack.pop().unwrap();
+                self.stack.pop().unwrap();
+            }
+            Instruction::BuildMap { size, .. } => {
+                let mut map = HashMap::new();
+                for _ in 0..size {
+                    let value = self.stack.pop().unwrap();
+                    let name = if let Some(StackValue::String(name)) = self.stack.pop() {
+                        name
+                    } else {
+                        unimplemented!("no string keys isn't yet supported in py_function!")
+                    };
+                    map.insert(name, value);
+                }
+                self.stack.push(StackValue::Map(map));
+            }
+            Instruction::MakeFunction => {
+                let name = if let Some(StackValue::String(name)) = self.stack.pop() {
+                    name
+                } else {
+                    panic!("Expected function name")
+                };
+                let code = if let Some(StackValue::Code(code)) = self.stack.pop() {
+                    code
+                } else {
+                    panic!("Expected function code")
+                };
+                let annotations = if let Some(StackValue::Map(map)) = self.stack.pop() {
+                    map
+                } else {
+                    panic!("Expected function annotations")
+                };
+                self.stack.push(StackValue::Function(Function {
+                    name,
+                    code,
+                    annotations,
+                }));
+            }
+            Instruction::Duplicate => {
+                let value = self.stack.last().unwrap().clone();
+                self.stack.push(value);
+            }
+            Instruction::Rotate { amount } => {
+                let mut values = Vec::new();
+
+                // Pop all values from stack:
+                for _ in 0..amount {
+                    values.push(self.stack.pop().unwrap());
```

Comment:
You might be able to do something like `values.extend(stack.drain(len-amount..).rev())` here